### ISSUE TITLE
BUG: flaky start-up failures with GHA builds

### ISSUE THREAD

**Issue by h-vetinari** [conda-forge/conda-smithy#2163]
The pytorch builds used to run fine on alma8, now with the switch to alma9, I'm seeing the following in https://github.com/conda-forge/pytorch-cpu-feedstock/pull/293
```
  + docker run --gpus all -v /home/runnerx/actions-runner/_work/pytorch-cpu-feedstock/pytorch-cpu-feedstock/recipe:/home/conda/recipe_root:rw,z,delegated -v /home/runnerx/actions-runner/_work/pytorch-cpu-feedstock/pytorch-cpu-feedstock:/home/conda/feedstock_root:rw,z,delegated -e CONFIG -e HOST_USER_ID -e UPLOAD_PACKAGES -e IS_PR_BUILD -e GIT_BRANCH -e UPLOAD_ON_BRANCH -e CI -e FEEDSTOCK_NAME -e CPU_COUNT -e BUILD_WITH_CONDA_DEBUG -e BUILD_OUTPUT_ID -e flow_run_id -e remote_url -e sha -e BINSTAR_TOKEN -e FEEDSTOCK_TOKEN -e STAGING_BINSTAR_TOKEN quay.io/condaforge/linux-anvil-x86_64:alma9 bash /home/conda/feedstock_root/.scripts/build_steps.sh
  docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: Auto-detected mode as 'legacy'
  nvidia-container-cli: initialization error: nvml error: driver not loaded: unknown.
  time="2024-11-23T21:12:58Z" level=error msg="error waiting for container: "
```

This is before we do any installation of our own CUDA packages, but more importantly, the same error occurs also for the non-CUDA-enabled build (which probably shouldn't be getting the `--gpus all`?). Also interesting is that native x64 builds are not affected by the same issue apparently.

CC @jakirkham @jaimergp @isuruf

**Comment by h-vetinari** [comment]
Nevermind, it now happened on x64 as well, and more importantly: on a job that previously [started](https://github.com/conda-forge/pytorch-cpu-feedstock/actions/runs/11990424477/job/33427923265) fine for the commit just before. So it's apparently a flaky thing.

**Comment by h-vetinari** [comment]
Aaaand now it even happened on the alma8 images (that were in use before already). So I'm guessing this might just be an issue for https://github.com/Quansight/open-gpu-server? I had opened here because I thought we might need to change the invocation for `docker run` to account for some image differences. Perhaps that's still the case?

**Comment by jaimergp** [comment]
@aktech, did anything change in the VM image recently? 🤔

**Comment by h-vetinari** [comment]
I don't understand the situation here. In [jaxlib](https://github.com/conda-forge/jaxlib-feedstock/actions/runs/11993479734/job/33439775312?pr=288), we also cross-compile linux-aarch, and there the builds haven't had an issue. OTOH, I've restarted the [pytorch](https://github.com/conda-forge/pytorch-cpu-feedstock/actions/runs/11990554768/job/33439563979?pr=293) job a number of times, and the linux-aarch64 + CUDA job has run into the same failure 5-6 times in a row, while all the other jobs ran through fine.

**Comment by aktech** [comment]
>@aktech, did anything change in the VM image recently? 🤔

Nothing has changed that I am aware of. Let me investigate to see if its indeed an infrastructure issue.

**Comment by aktech** [comment]
Is there a minimally reproducible example that I can use?

Something like the following maybe?

```bash
docker run --gpus all quay.io/condaforge/linux-anvil-x86_64:alma9 nvidia-smi
```

**Comment by h-vetinari** [comment]
@aktech, have a look here: https://github.com/conda-forge/pytorch-cpu-feedstock/pull/294 
It fails about 1 in every 6 runs.

**Comment by h-vetinari** [comment]
So it turns out the 1-in-6 rate was unusually lucky. In the other pytorch PR, we're getting 3-4 out of 6 runs that fail, requiring multiple restarts, which makes the cirun server barely usable for us.

Have you been able to find anything in your investigation so far @aktech?

**Comment by aktech** [comment]
Hey @h-vetinari 

I tried debugging and found out this:

## nvidia-smi
```bash
$ nvidia-smi
NVIDIA-SMI has failed because it couldn't communicate with the NVIDIA driver. Make sure that the latest NVIDIA driver is installed and running.
```

## GPU attached to VM

```
$ lspci | grep NVIDIA
00:05.0 3D controller: NVIDIA Corporation GV100GL [Tesla V100 PCIe 16GB] (rev a1)
```

## Drivers installed

```
$ dpkg -l | grep nvidia-driver
hi  nvidia-driver-525                     525.125.06-0ubuntu0.22.04.1             amd64        NVIDIA driver metapackage
...
```

Nvidia GPU is attached to the VM, drivers are installed, but nvidia-smi is not seeing it for some reason.

Here are the kernel logs:

```
[ 1249.381189] NVRM: The NVIDIA probe routine failed for 1 device(s).
[ 1249.381192] NVRM: None of the NVIDIA devices were initialized.
[ 1249.514598] NVRM: The NVIDIA GPU 0000:00:05.0
               NVRM: (PCI ID: 10de:1db4) installed in this system has
               NVRM: fallen off the bus and is not responding to commands.
[ 1249.565014] NVRM: The NVIDIA probe routine failed for 1 device(s).
[ 1249.565016] NVRM: None of the NVIDIA devices were initialized.
```

The final potential solution is to try updating the image to install latest drivers and see if that fixes the problem, but before that I would like to get to the root cause. I'll get back with more details soon (assuming the GPU is functioning physically and there is no low power supply issue).


>requiring multiple restarts, which makes the cirun server barely usable for us.

Apologies for the inconvenience on this. Also, a clarification, this is open gpu server and not cirun server, cirun is just a service to orchestrate runners :)

**Comment by h-vetinari** [comment]
Thank you for your help! I think we could definitely try to update the drivers in the image. Though that scenario wouldn't really explain the fact that it sometimes works and sometimes doesn't, because the image is always the same... 🤔

> Also, a clarification, this is open gpu server and not cirun server, cirun is just a service to orchestrate runners :)

Yeah, I'm aware. I'm just grasping at a good name honestly. In "the XYZ server", I think XYZ=cirun works reasonably well. "Open GPU" sounds like a huge initiative, but I just want to refer to this one particuliar server.

**Comment by aktech** [comment]
>Though that scenario wouldn't really explain the fact that it sometimes works and sometimes doesn't, because the image is always the same... 🤔

Yes exactly. Trying to figure that out. 🤞

**Comment by aktech** [comment]
@h-vetinari 

I suspect hardware issues, looking at the kernel log.

Most likely thermal or PSU issue, might need to talk to someone in the datacenter (I have started the dialog). The above-mentioned error happens on a particular pair of GPUs from what I can tell (I have disabled the suspected GPUs for now). Feel free to trigger jobs now.

Ref:
- https://forums.developer.nvidia.com/t/gpu-has-fallen-off-the-bus/293334/2
- https://forums.developer.nvidia.com/t/gpu-has-fallen-off-the-bus/217357/4
